src/lazy/binary/raw/v1_1/value.rs

```diff
@@ -95,31 +105,104 @@ impl<'top> HasSpan<'top> for LazyRawBinaryValue_1_1<'top> {
+            } else {
+                IonType::Null
+            };
+            return Ok(RawValueRef::Null(ion_type));
+        }
+
+        match self.ion_type() {
+            IonType::Null => unreachable!("all null types handled above"),
+            IonType::Bool => Ok(RawValueRef::Bool(self.read_bool()?)),
+            IonType::Int => Ok(RawValueRef::Int(self.read_int()?)),
+            IonType::Float => Ok(RawValueRef::Float(self.read_float()?)),
+            IonType::Decimal => Ok(RawValueRef::Decimal(self.read_decimal()?)),
+            IonType::Timestamp => Ok(RawValueRef::Timestamp(self.read_timestamp()?)),
+            IonType::Symbol => Ok(RawValueRef::Symbol(self.read_symbol()?)),
+            IonType::String => Ok(RawValueRef::String(self.read_string()?)),
+            IonType::Clob => Ok(RawValueRef::Clob(self.read_clob()?)),
+            IonType::Blob => Ok(RawValueRef::Blob(self.read_blob()?)),
+            IonType::List => Ok(RawValueRef::List(self.read_list()?)),
+            IonType::SExp => Ok(RawValueRef::SExp(self.read_sexp()?)),
+            IonType::Struct => Ok(RawValueRef::Struct(self.read_struct()?)),
+        }
+    }
+
+    /// This is a fast path for reading values that we know need to be resolved.
+    ///
+    /// When a `LazyValue` wrapping a raw binary value calls `read()`, it's clear that the `RawValueRef` will
+    /// need to be resolved into a `ValueRef` before it is returned to the application. If `LazyValue::read`
+    /// (indirectly) calls `RawLazyValue::read`, the raw level will dispatch on the Ion type of the current
+    /// value to find the correct decoder. It will then return the `RawValueRef` to the `LazyValue`, which
+    /// will dispatch on the Ion type again to resolve it into a `ValueRef`. These two identical dispatch steps
+    /// over 13 Ion types happen far enough away from each other in the code that they cannot be consolidated
+    /// into a single dispatch by the compiler.
+    ///
+    /// This method exists to perform the read and resolve steps in the same locale, allowing the compiler
+    /// to optimize it more effectively.
+    #[inline(always)]
+    fn read_resolved(
+        &self,
+        context: EncodingContextRef<'top>,
+    ) -> IonResult<ValueRef<'top, BinaryEncoding_1_1>> {
+        if self.is_null() {
+            return Ok(ValueRef::Null(self.ion_type()));
+        }
+        // Anecdotally, string and integer values are very common in Ion streams. This `match` creates
+        // an inlineable fast path for them while other types go through the general case impl.
+        // NOTE: We can and should change the subset of types this optimizes for when we have data to
+        //       better inform our decision.
+        match self.ion_type() {
+            IonType::String => Ok(ValueRef::String(self.read_string()?)),
+            IonType::Int => Ok(ValueRef::Int(self.read_int()?)),
+            _ => self.read_resolved_general_case(context),
```

Comment:
> I don't think it's too long to put inline in read_resolved

Sorry, what's not too long?

`read_resolved_general_case` handles _all_ possibilities (including string and int). This is only creating a more-likely-to-be-inlined path for strings and ints.

Reply:
> Sorry, what's not too long?

Well...

> `read_resolved_general_case` is defined in an inherent impl block further below. In retrospect, I could have defined it nested inside `read_resolved`. What do you think?

I don't think the definition of `read_resolved_general_case` is too long to define nested inside `read_resolved`—i.e. it won't hurt the readability of the code to do so. I also think that (assuming there are no terrible performance implications) it would be fine to put all of the branches right here rather than having them defined in a separate function.